src/sentry/migrations/0361_break_project_integration_fk.py

```diff
@@ -0,0 +1,57 @@
+# Generated by Django 2.2.28 on 2023-02-22 21:05
+
+from django.db import migrations
+
+import sentry.db.models.fields.hybrid_cloud_foreign_key
+from sentry.new_migrations.migrations import CheckedMigration
+
+
+class Migration(CheckedMigration):
+    # This flag is used to mark that a migration shouldn't be automatically run in production. For
+    # the most part, this should only be used for operations where it's safe to run the migration
+    # after your code has deployed. So this should not be used for most operations that alter the
+    # schema of a table.
+    # Here are some things that make sense to mark as dangerous:
+    # - Large data migrations. Typically we want these to be run manually by ops so that they can
+    #   be monitored and not block the deploy for a long period of time while they run.
+    # - Adding indexes to large tables. Since this can take a long time, we'd generally prefer to
+    #   have ops run this and not block the deploy. Note that while adding an index is a schema
+    #   change, it's completely safe to run the operation after the code has deployed.
+    is_dangerous = False
+
+    dependencies = [
+        ("sentry", "0360_authenticator_config_type_change"),
+    ]
+
+    database_operations = [
+        migrations.AlterField(
+            model_name="projectintegration",
+            name="integration",
+            field=sentry.db.models.fields.foreignkey.FlexibleForeignKey(
+                to="sentry.Integration", db_constraint=False, db_index=True
+            ),
+        ),
+    ]
+
+    state_operations = [
+        migrations.AlterField(
+            model_name="projectintegration",
+            name="integration",
+            field=sentry.db.models.fields.hybrid_cloud_foreign_key.HybridCloudForeignKey(
+                "sentry.Integration", db_index=True, on_delete="CASCADE"
+            ),
+        ),
+        migrations.RenameField(
+            model_name="projectintegration",
+            old_name="integration",
+            new_name="integration_id",
+        ),
+        migrations.AlterUniqueTogether(
+            name="projectintegration",
+            unique_together={("project", "integration_id")},
+        ),
+    ]
+
+    operations = database_operations + [
+        migrations.SeparateDatabaseAndState(state_operations=state_operations)
```

Comment:
Yup.  So this is the pattern I have to use for all these breaking fk migations and its intentional.

Runnning it both as DDL and state change is fine because the subsequent state operations will set the state correctly in a transaction.

The reason I have to do this is -- for "mysterious reasons" django migration tooling will only produce SQL if I do so.  If I attach the database operations inside of the `SeperateDatabaseAndState` migration, it runs, but it doesn't produce SQL, thus failing CI.

🤷

Reply:
>  If I attach the database operations inside of the SeperateDatabaseAndState migration, it runs, but it doesn't produce SQL, thus failing CI.

Interesting that django can't figure out what the SQL it is going to generate will be. Good to know that what you have is intentional for these migrations. I'll stop commenting on them :smile: